### Pacotes utilizados

In [115]:
import pandas as pd
import numpy as np

### Base utilizada

In [116]:
df_data = pd.read_csv('./Base de dados/ConsumidorGov.csv',sep = ';')

# Atribuindo Dados

## 1. Atribuindo constantes

Ao fazer a atribuição de uma coluna de um DataFrame, como por exemplo á coluna `Cidade` do DataFrame `df_data`

In [117]:
df_data.head()

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   
3  Superintendência de Proteção e Defesa do Consu...  Plataforma Web     NE   
4        Fundação de Proteção e Defesa do Consumidor  Plataforma Web     SE   

   UF    Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL    Maceió    F  entre 61 a 70 anos          2024            10   
1  AL    Maceió    F  entre 61 a 70 anos          2024            10   
2  PE    Olinda    F  entre 41 a 50 anos          2024            10   
3  BA  Salvador    M  entre 21 a 30 anos          2024            10   
4  SP   Guarujá    M  entre 51 a 60 anos          2024            10   

  Data Abertura  ...                                            Assunto  \
0    01/10/2024  ...                               Atendimento Bancário   
1    01/10/2024  ...                               Atendimento Bancário   
2    01/10/2024  ...                    Aplicativo de mobilidade urbana   
3    01/10/2024  ...  Cartão de Crédito / Cartão de Débito / Cartão ...   
4    01/10/2024  ...                                      Água e Esgoto   

           Grupo Problema                                           Problema  \
0       Atendimento / SAC                Descumprimento do estatuto do idoso   
1       Atendimento / SAC                Descumprimento do estatuto do idoso   
2  Cobrança / Contestação  Cobrança por serviço não fornecido/ em desacor...   
3              Informação  Informações incompletas / inadequadas sobre o ...   
4       Atendimento / SAC  Má qualidade no atendimento (descortesia / des...   

  Como Comprou Contratou Procurou Empresa  Respondida             Situação  \
0               Telefone                S           S  Finalizada avaliada   
1               Telefone                S           S  Finalizada avaliada   
2               Internet                S           S  Finalizada avaliada   
3               Internet                S           S  Finalizada avaliada   
4            Loja física                S           S  Finalizada avaliada   

  Avaliação Reclamação Nota do Consumidor Análise da Recusa  
0            Resolvida                5.0               NaN  
1            Resolvida                5.0               NaN  
2        Não Resolvida                1.0               NaN  
3            Resolvida                5.0               NaN  
4            Resolvida                3.0               NaN  

[5 rows x 30 columns]

In [118]:
df_data['Cidade']

0           Maceió
1           Maceió
2           Olinda
3         Salvador
4          Guarujá
            ...   
105013     Ibirité
105014     Ibirité
105015     Ibirité
105016     Ibirité
105017     Ibirité
Name: Cidade, Length: 105018, dtype: object

In [119]:
cidade_view = df_data['Cidade']
cidade_view

0           Maceió
1           Maceió
2           Olinda
3         Salvador
4          Guarujá
            ...   
105013     Ibirité
105014     Ibirité
105015     Ibirité
105016     Ibirité
105017     Ibirité
Name: Cidade, Length: 105018, dtype: object

Esta **NÃO É UMA CÓPIA**, mas sim, uma **REFERÊNCIA** à coluna do DataFrame

Para fazer de fato uma cópia e criar uma Series/DataFrame que contenha não apenas uma referência, é necessário utilizar o método `copy()`

In [120]:
cidade_copy = df_data['Cidade'].copy()
cidade_copy

0           Maceió
1           Maceió
2           Olinda
3         Salvador
4          Guarujá
            ...   
105013     Ibirité
105014     Ibirité
105015     Ibirité
105016     Ibirité
105017     Ibirité
Name: Cidade, Length: 105018, dtype: object

E qual a diferença entre **REFERÊNCIA** e **CÓPIA**?
Caso alguma alteração no DataFrame original `df_data` seja feita, temos que:

Para exemplificar utilizando a função `iloc`, esta nos permite selecionar linhas e colunas no DataFrame com base em uma posição númerica, vamos alterar então a **LINHA 0** e **COLUNA 4**, para isso podemos fazer da seguinte forma:

In [121]:
df_data.iloc[0,4] = 'Alterado'

In [122]:
df_data.head(1)

Gestor Canal de Origem Região  UF    Cidade  \
0  Secretaria Nacional do Consumidor  Plataforma Web     NE  AL  Alterado   

  Sexo        Faixa Etária  Ano Abertura  Mês Abertura Data Abertura  ...  \
0    F  entre 61 a 70 anos          2024            10    01/10/2024  ...   

                Assunto     Grupo Problema  \
0  Atendimento Bancário  Atendimento / SAC   

                              Problema Como Comprou Contratou  \
0  Descumprimento do estatuto do idoso               Telefone   

  Procurou Empresa  Respondida             Situação Avaliação Reclamação  \
0                S           S  Finalizada avaliada            Resolvida   

  Nota do Consumidor Análise da Recusa  
0                5.0               NaN  

[1 rows x 30 columns]

Após a alteração, se buscarmos a **CÓPIA** teremos o seguinte:

In [123]:
cidade_copy

0           Maceió
1           Maceió
2           Olinda
3         Salvador
4          Guarujá
            ...   
105013     Ibirité
105014     Ibirité
105015     Ibirité
105016     Ibirité
105017     Ibirité
Name: Cidade, Length: 105018, dtype: object

Após a alteração, se buscarmos a **REFERÊNCIA** teremos o seguinte:

In [124]:
cidade_view

0         Alterado
1           Maceió
2           Olinda
3         Salvador
4          Guarujá
            ...   
105013     Ibirité
105014     Ibirité
105015     Ibirité
105016     Ibirité
105017     Ibirité
Name: Cidade, Length: 105018, dtype: object

Nota-se então, que os valores da **CÓPIA** se mantém mesmo após a alteração do DataFrame original `df_data`, enquanto a **REFERÊNCIA** por ser um ponteiro ao DataFrame original também tem seu valor alterado.

A escolha entre view (referência direta) e cópia (duplicação dos dados) no pandas depende do contexto em que você está trabalhando e dos objetivos da operação. Cada abordagem tem vantagens e desvantagens e deve ser usada com cuidado para evitar comportamentos inesperados ou problemas de desempenho.

## 2. Atribuindo listas ou series

In [125]:
n_linhas, n_colum = df_data.shape
n_linhas, n_colum

(105018, 30)

O nosso DataFram `df_data` contém 105018 linhas e 30 colunas, para fazer a alteração de uma coluna inteira precisamos de uma **Series** contendo o mesmo número de linhas existente na coluna objetivo ou seja, uma lista contendo 105018

Para criação destes 105018 valores de uma coluna utilizaremos **LIST COMPREHENSION**, que é uma forma de criar listas a partir de outras listas, usando uma notação matemática.

In [126]:
novas_cidades = [f'Cidade {i}'for i in range(n_linhas)]
len(novas_cidades) #Retorna o número de itens de um objeto, como uma lista, string, tupla, dicionário ou qualquer sequência

105018

A quantidade de elementos na lista `novas_cidades` é igual ao número de linhas do DataFrame `df_data`, podemos então fazer a substituição dos valores da coluna **Cidade** para o novo valor

In [127]:
df_data['Cidade'] = novas_cidades

In [128]:
df_data.head()

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   
3  Superintendência de Proteção e Defesa do Consu...  Plataforma Web     NE   
4        Fundação de Proteção e Defesa do Consumidor  Plataforma Web     SE   

   UF    Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Cidade 0    F  entre 61 a 70 anos          2024            10   
1  AL  Cidade 1    F  entre 61 a 70 anos          2024            10   
2  PE  Cidade 2    F  entre 41 a 50 anos          2024            10   
3  BA  Cidade 3    M  entre 21 a 30 anos          2024            10   
4  SP  Cidade 4    M  entre 51 a 60 anos          2024            10   

  Data Abertura  ...                                            Assunto  \
0    01/10/2024  ...                               Atendimento Bancário   
1    01/10/2024  ...                               Atendimento Bancário   
2    01/10/2024  ...                    Aplicativo de mobilidade urbana   
3    01/10/2024  ...  Cartão de Crédito / Cartão de Débito / Cartão ...   
4    01/10/2024  ...                                      Água e Esgoto   

           Grupo Problema                                           Problema  \
0       Atendimento / SAC                Descumprimento do estatuto do idoso   
1       Atendimento / SAC                Descumprimento do estatuto do idoso   
2  Cobrança / Contestação  Cobrança por serviço não fornecido/ em desacor...   
3              Informação  Informações incompletas / inadequadas sobre o ...   
4       Atendimento / SAC  Má qualidade no atendimento (descortesia / des...   

  Como Comprou Contratou Procurou Empresa  Respondida             Situação  \
0               Telefone                S           S  Finalizada avaliada   
1               Telefone                S           S  Finalizada avaliada   
2               Internet                S           S  Finalizada avaliada   
3               Internet                S           S  Finalizada avaliada   
4            Loja física                S           S  Finalizada avaliada   

  Avaliação Reclamação Nota do Consumidor Análise da Recusa  
0            Resolvida                5.0               NaN  
1            Resolvida                5.0               NaN  
2        Não Resolvida                1.0               NaN  
3            Resolvida                5.0               NaN  
4            Resolvida                3.0               NaN  

[5 rows x 30 columns]

Para retornarmos os dados de **Cidade** originais, podemos utilizar a cópia feita na Series `cidade_copy`, esta Series foi criada e copiada antes da alteração da coluna, portanto tem os valores originais da mesma

In [129]:
df_data['Cidade'] = cidade_copy

In [130]:
df_data.head()

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   
3  Superintendência de Proteção e Defesa do Consu...  Plataforma Web     NE   
4        Fundação de Proteção e Defesa do Consumidor  Plataforma Web     SE   

   UF    Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL    Maceió    F  entre 61 a 70 anos          2024            10   
1  AL    Maceió    F  entre 61 a 70 anos          2024            10   
2  PE    Olinda    F  entre 41 a 50 anos          2024            10   
3  BA  Salvador    M  entre 21 a 30 anos          2024            10   
4  SP   Guarujá    M  entre 51 a 60 anos          2024            10   

  Data Abertura  ...                                            Assunto  \
0    01/10/2024  ...                               Atendimento Bancário   
1    01/10/2024  ...                               Atendimento Bancário   
2    01/10/2024  ...                    Aplicativo de mobilidade urbana   
3    01/10/2024  ...  Cartão de Crédito / Cartão de Débito / Cartão ...   
4    01/10/2024  ...                                      Água e Esgoto   

           Grupo Problema                                           Problema  \
0       Atendimento / SAC                Descumprimento do estatuto do idoso   
1       Atendimento / SAC                Descumprimento do estatuto do idoso   
2  Cobrança / Contestação  Cobrança por serviço não fornecido/ em desacor...   
3              Informação  Informações incompletas / inadequadas sobre o ...   
4       Atendimento / SAC  Má qualidade no atendimento (descortesia / des...   

  Como Comprou Contratou Procurou Empresa  Respondida             Situação  \
0               Telefone                S           S  Finalizada avaliada   
1               Telefone                S           S  Finalizada avaliada   
2               Internet                S           S  Finalizada avaliada   
3               Internet                S           S  Finalizada avaliada   
4            Loja física                S           S  Finalizada avaliada   

  Avaliação Reclamação Nota do Consumidor Análise da Recusa  
0            Resolvida                5.0               NaN  
1            Resolvida                5.0               NaN  
2        Não Resolvida                1.0               NaN  
3            Resolvida                5.0               NaN  
4            Resolvida                3.0               NaN  

[5 rows x 30 columns]

## 3. Criando novas colunas

Para se criar uma **nova coluna** em um DataFrame é necessário atribuirmos uma lista/Series de valores ou uma constante a uma nova chave do DataFrame. Da mesmo forma da atribuição, a quantidade de valores da lista precisa ser igual ao número de linhas do DataFrame.

### 3.1 Criando colunas a partir de valores constantes

Para criar colunas a partir de um valor constante, ou seja, todas as linhas terão o mesmo valor para esta nova coluna

In [131]:
df_data['Coluna Constante'] = 'Valor Constante'
df_data.head(3)

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   

   UF  Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Maceió    F  entre 61 a 70 anos          2024            10   
1  AL  Maceió    F  entre 61 a 70 anos          2024            10   
2  PE  Olinda    F  entre 41 a 50 anos          2024            10   

  Data Abertura  ...          Grupo Problema  \
0    01/10/2024  ...       Atendimento / SAC   
1    01/10/2024  ...       Atendimento / SAC   
2    01/10/2024  ...  Cobrança / Contestação   

                                            Problema Como Comprou Contratou  \
0                Descumprimento do estatuto do idoso               Telefone   
1                Descumprimento do estatuto do idoso               Telefone   
2  Cobrança por serviço não fornecido/ em desacor...               Internet   

  Procurou Empresa Respondida             Situação  Avaliação Reclamação  \
0                S          S  Finalizada avaliada             Resolvida   
1                S          S  Finalizada avaliada             Resolvida   
2                S          S  Finalizada avaliada         Não Resolvida   

  Nota do Consumidor Análise da Recusa Coluna Constante  
0                5.0               NaN  Valor Constante  
1                5.0               NaN  Valor Constante  
2                1.0               NaN  Valor Constante  

[3 rows x 31 columns]

Agora, a nova coluna `Coluna Constante` foi criada e ela contém em todas as 105018 linhas o mesmo valor, neste caso a string **'Constante'**

### 3.2 Criando colunas a partir de uma lista

Para criar colunas a partir de uma lista pré definida, basta atribuir ao DataFrame o novo campo com a lista

In [132]:
df_data['Coluna Lista'] = range(df_data.shape[0])
df_data.head(3)

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   

   UF  Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Maceió    F  entre 61 a 70 anos          2024            10   
1  AL  Maceió    F  entre 61 a 70 anos          2024            10   
2  PE  Olinda    F  entre 41 a 50 anos          2024            10   

  Data Abertura  ...                                           Problema  \
0    01/10/2024  ...                Descumprimento do estatuto do idoso   
1    01/10/2024  ...                Descumprimento do estatuto do idoso   
2    01/10/2024  ...  Cobrança por serviço não fornecido/ em desacor...   

  Como Comprou Contratou Procurou Empresa Respondida             Situação  \
0               Telefone                S          S  Finalizada avaliada   
1               Telefone                S          S  Finalizada avaliada   
2               Internet                S          S  Finalizada avaliada   

   Avaliação Reclamação  Nota do Consumidor Análise da Recusa  \
0             Resolvida                 5.0               NaN   
1             Resolvida                 5.0               NaN   
2         Não Resolvida                 1.0               NaN   

  Coluna Constante Coluna Lista  
0  Valor Constante            0  
1  Valor Constante            1  
2  Valor Constante            2  

[3 rows x 32 columns]

Agora, a nova coluna `Coluna Lista` foi criada e ela contém em todas as 105018, neste caso o intervalo entre número 0 até 105017

Se tentarmos atribuir uma lista menor do que a quantidade de linhas totais do DataFrame teremos um problema de **Match**. Exemplo, se tentarmos atribuir uma lista de 3 itens no DataFrame `df_data`:

In [133]:
df_data['Lista Menor'] = [1,2,3]

ValueError: Length of values (3) does not match length of index (105018)

### 3.3 Criando colunas a partir de outra coluna

Uma coluna pode ser criado com base em calculos matemátios e/ou condicionais dependentes de outros campos. Por exemplo, se tentarmos criar uma nova coluna chamada `Nova Nota Consumidor` onde o valor dela é a multiplicação do campo `Nota Consumidor` * **2**, fariamos o seguintes:

In [134]:
df_data['Nova Nota do Consumidor'] = df_data['Nota do Consumidor'] * 2
df_data.head(3)

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   

   UF  Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Maceió    F  entre 61 a 70 anos          2024            10   
1  AL  Maceió    F  entre 61 a 70 anos          2024            10   
2  PE  Olinda    F  entre 41 a 50 anos          2024            10   

  Data Abertura  ... Como Comprou Contratou Procurou Empresa Respondida  \
0    01/10/2024  ...               Telefone                S          S   
1    01/10/2024  ...               Telefone                S          S   
2    01/10/2024  ...               Internet                S          S   

              Situação Avaliação Reclamação  Nota do Consumidor  \
0  Finalizada avaliada            Resolvida                 5.0   
1  Finalizada avaliada            Resolvida                 5.0   
2  Finalizada avaliada        Não Resolvida                 1.0   

   Análise da Recusa Coluna Constante Coluna Lista Nova Nota do Consumidor  
0                NaN  Valor Constante            0                    10.0  
1                NaN  Valor Constante            1                    10.0  
2                NaN  Valor Constante            2                     2.0  

[3 rows x 33 columns]

### 3.4 Criando colunas condicionais

In [135]:
df_data['Coluna Condicional'] = np.where(df_data['Nova Nota do Consumidor'] >= 5, 'Boa','Ruim')
df_data.head(3)

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   

   UF  Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Maceió    F  entre 61 a 70 anos          2024            10   
1  AL  Maceió    F  entre 61 a 70 anos          2024            10   
2  PE  Olinda    F  entre 41 a 50 anos          2024            10   

  Data Abertura  ... Procurou Empresa Respondida             Situação  \
0    01/10/2024  ...                S          S  Finalizada avaliada   
1    01/10/2024  ...                S          S  Finalizada avaliada   
2    01/10/2024  ...                S          S  Finalizada avaliada   

  Avaliação Reclamação Nota do Consumidor  Análise da Recusa  \
0            Resolvida                5.0                NaN   
1            Resolvida                5.0                NaN   
2        Não Resolvida                1.0                NaN   

   Coluna Constante Coluna Lista Nova Nota do Consumidor Coluna Condicional  
0   Valor Constante            0                    10.0                Boa  
1   Valor Constante            1                    10.0                Boa  
2   Valor Constante            2                     2.0               Ruim  

[3 rows x 34 columns]

### 3.4 Dropando campos

Para fazer o drop de campos que não serão utilizados no DataFrame, basta utilizar o método `drop()`, da seguinte forma:

In [136]:
df_data.drop(['Coluna Constante','Coluna Lista','Nova Nota do Consumidor','Coluna Condicional'], axis=1, inplace=True)
df_data.head(3)

Gestor Canal de Origem Região  \
0                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
1                  Secretaria Nacional do Consumidor  Plataforma Web     NE   
2  Secretaria de Justiça e Direitos Humanos - Sec...  Plataforma Web     NE   

   UF  Cidade Sexo        Faixa Etária  Ano Abertura  Mês Abertura  \
0  AL  Maceió    F  entre 61 a 70 anos          2024            10   
1  AL  Maceió    F  entre 61 a 70 anos          2024            10   
2  PE  Olinda    F  entre 41 a 50 anos          2024            10   

  Data Abertura  ...                          Assunto          Grupo Problema  \
0    01/10/2024  ...             Atendimento Bancário       Atendimento / SAC   
1    01/10/2024  ...             Atendimento Bancário       Atendimento / SAC   
2    01/10/2024  ...  Aplicativo de mobilidade urbana  Cobrança / Contestação   

                                            Problema Como Comprou Contratou  \
0                Descumprimento do estatuto do idoso               Telefone   
1                Descumprimento do estatuto do idoso               Telefone   
2  Cobrança por serviço não fornecido/ em desacor...               Internet   

  Procurou Empresa  Respondida             Situação Avaliação Reclamação  \
0                S           S  Finalizada avaliada            Resolvida   
1                S           S  Finalizada avaliada            Resolvida   
2                S           S  Finalizada avaliada        Não Resolvida   

  Nota do Consumidor Análise da Recusa  
0                5.0               NaN  
1                5.0               NaN  
2                1.0               NaN  

[3 rows x 30 columns]